### ISSUE TITLE
Enable PBClientFactory and PBServerFactory to use custom security options.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1879]
|<img alt="pgstudy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| pgstudy reported|
|-|-|
|Trac ID|trac#1879|
|Type|enhancement|
|Created|2006-06-30 00:24:44Z|

Sometimes we wish to have custom security options, i.e., decide ourselves what Modules or Classes to receive. This can be done using Broker, however, PBClientFactory and PBServerFactory can have this feature as well.
This is just a few lines of changes.

Attachments:

* [CSFactorySecurity.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/143/143df602a55fa43b4454d778d1fe8555f9d162d0/6537ae76bd6deda2a2e68ae75a24d0b4872022c9.diff) (1331 bytes) - added by pgstudy on 2006-06-30 00:27:11Z - The patch
* [testnis.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/143/143df602a55fa43b4454d778d1fe8555f9d162d0/ba71e1b31b7da0e372c15469bc6c1e1edb1a5b3e.py) (2190 bytes) - added by pgstudy on 2006-06-30 00:30:51Z - test for patch
* [test_pbsec.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/143/143df602a55fa43b4454d778d1fe8555f9d162d0/91ad924cf30d3266e1fd8ce317958f388385ba1e.py) (803 bytes) - added by pgstudy on 2006-07-01 17:11:00Z - Test without reactor.run()
* [CSFactorySecurity.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/143/143df602a55fa43b4454d778d1fe8555f9d162d0/5f482b1c81677a0c08a7c26844a618a6678656a7.diff) (1351 bytes) - added by pgstudy on 2006-07-01 18:07:09Z - security are instantiate at Broker for ServerFacotry

<details><summary>Searchable metadata</summary>

```
trac-id__1879 1879
type__enhancement enhancement
reporter__pgstudy pgstudy
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1151627084000000 1151627084000000
changetime__1194298115000000 1194298115000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to pgstudy|
|-|-|

Tests shouldn't have prints in them.  If there really is very important information which needs to be recorded for someone to analyse after the test run, use log.msg().

The test should be using twisted.trial.unittest instead of the stdlib unittest.  This will let it....


Return a Deferred from the test method.  _Don't_ call reactor.run() or reactor.stop() in the test.  But...

You don't actually need to set up a connection here.  You just need to make sure that the broker is instantiated with the right security options.  You can do this by directly calling buildProtocol on PBServerFactory and then inspecting the resulting object.

There should be a docstring explaining the purpose of the test.

The Twisted coding standard is here - <http://twistedmatrix.com/projects/core/documentation/howto/policy/coding-standard.html>.  It isn't completely exhaustive, so also please take a look at some other code in Twisted and try to make yours look similar.

**Comment by twisted-trac** [comment]
|<img alt="pgstudy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|pgstudy commented|
|-|-|

However, you will note that i also changed the PBClientFactory security settings and thsoe are only assigned when the function PBClientFactory.clientConnectionMade is executed which i guess only happens when the connection is made :)
I.e., how can i not call reactor run and stop?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

One possibility is to change the code.  One mantra of XP is "if it is hard to test it is wrong".  Another option is to use a loopback connection.  The first option seems preferable to me: is there any strong reason not to assign the security options in buildProtocol?

**Comment by twisted-trac** [comment]
|<img alt="pgstudy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|pgstudy commented|
|-|-|

exarkun: For the PBServerFactory, the buildprotocol is where the Broker is instantiated. I don't know of any other suitable place. For PBClientFactory, the entry point for the broker is the connectionMade, again, i don't see any other suitable place.

What is a loopback connection, currently i connect to the local host. Point me to the suitable test example that uses loopback :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Why can't the Broker be created in the same method which sets its security options?

There are tons of examples of loopback connections in Twisted's test suite.  Just grep for 'lookback' in test_*.py and you should find plenty.

**Comment by twisted-trac** [comment]
|<img alt="pgstudy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|pgstudy commented|
|-|-|

ok, at least for the ServerFactory i changed the patch to set the security options when instatiating in the buildprotocol. In the client factory, the Broker is handed out from outside the clientfactory so seemingly it must be set directly.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [21323]) Add the security parameter to client and server factories, test it.

Refs [#1879](https://github.com/twisted/twisted/issues/1879)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @therve|
|-|-|

This is done in pb-security-options-1879, waiting a bit.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

```None``` is a better dummy than ```""``` I think - the four new tests pass the latter to ```buildProtocol```.

The change to PBServerFactory's docstring to use ```C{} ```would be trivially better if it used ```L{} ```instead.

In the docstring for PBClientFactory.!__init!__, I think ```tracebacks for exceptions``` reads slightly better than ```tracebacks of exceptions```.  Same for PBServerFactory.!__init!__'s docstring.

There's a typo in the ```@type``` for ```security``` in both of those docstrings, too, ```spred```.  That could also be using ```L{} ```.

Another typo in the docstring for PBClientFactory.buildProtocol - ```broket```.

The documentation for the ```root``` parameter to PBServerFactory is a bit misleading, since root isn't really the root Referenceable, but a factory which will provide the root reference (via its rootObject method).

So lots of nits.  The actual functionality looks great.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

(In [21541]) Process review comments.

Refs [#1879](https://github.com/twisted/twisted/issues/1879)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve removed owner|
|-|-|

Thanks for your rigor :). Back to you.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @therve|
|-|-|

Any reason PBServerFactory.!__init!__.!__doc!__ still has ```C{IPBRoot} ```and ```C{twisted.spread.jelly.SecurityOptions} ```instead of using ```L{} ```for each?

The rest looks great.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

(In [21584]) Forgot some stuff.

Refs [#1879](https://github.com/twisted/twisted/issues/1879)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set status to `closed`|
|-|-|

(In [21593]) Merge pb-security-options-1879

Author: therve, pgstudy
Reviewer: exarkun
Fixes [#1879](https://github.com/twisted/twisted/issues/1879)

Enable pb factories to use custom security options.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
